### PR TITLE
Stop returning overlapping ranges from TimeWindowPartitionMapping 

### PR DESCRIPTION
**Pull request by clairelin135**
When TimeWindowPartitionMapping._map_partitions is called with an input partitions subset containing multiple ranges, it can return a partitions subset that contains overlapping ranges. This happens when TimeWindowPartitionMapping's start offset or end offset is non-zero.

This PR fixes -- thanks @OwenKephart

### CODE REVIEW

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* **#20112** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20112?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/20112?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py [hunk 1]
```diff
@@ -161,6 +161,26 @@ def get_downstream_partitions_for_partitions(
...
+
+        sorted_time_windows = sorted(time_windows, key=lambda tw: tw.start.timestamp())
+        merged_time_windows: List[TimeWindow] = [sorted_time_windows[0]]
+
+        for window in sorted_time_windows[1:]:
+            last_window = merged_time_windows[-1]
+            # window starts before the end of the previous one
+            if window.start.timestamp() <= last_window.end.timestamp():
+                merged_time_windows[-1] = TimeWindow(
+                    last_window.start, max(last_window.end, window.end, key=lambda d: d.timestamp())
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py:177]: these timestamp() calls are unfortunately necessary even though datetimes support comparison operations due to the cursed knowledge that these comparison operators sometimes return the wrong answer when dealing with DST transitions :|

Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py:177]: dst strikes again

**Review by clairelin135 — COMMENTED** [review]
Looks great to me, though I can't approve because I'm the original creator of the PR...

Thanks for fixing!

**Review by OwenKephart — APPROVED** [review]

**Pull request merged**